source/ocproute.go

```diff
@@ -0,0 +1,281 @@
+	namespace string,
+	annotationFilter string,
+	fqdnTemplate string,
+	combineFQDNAnnotation bool,
+	ignoreHostnameAnnotation bool,
+) (Source, error) {
+	var (
+		tmpl *template.Template
+		err  error
+	)
+	if fqdnTemplate != "" {
+		tmpl, err = template.New("endpoint").Funcs(template.FuncMap{
+			"trimPrefix": strings.TrimPrefix,
+		}).Parse(fqdnTemplate)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Use shared informer to listen for add/update/delete of Routes in the specified namespace.
+	// Set resync period to 0, to prevent processing when nothing has changed.
+	informerFactory := extInformers.NewFilteredSharedInformerFactory(ocpClient, 0, namespace, nil)
+	routeInformer := informerFactory.Route().V1().Routes()
+
+	// Add default resource event handlers to properly initialize informer.
+	routeInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    func(obj interface{}) {
+			},
+		},
+	)
+
+	// TODO informer is not explicitly stopped since controller is not passing in its channel.
+	informerFactory.Start(wait.NeverStop)
+
+	// wait for the local cache to be populated.
+	err = wait.Poll(time.Second, 60*time.Second, func() (bool, error) {
+		return routeInformer.Informer().HasSynced(), nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to sync cache: %v", err)
+	}
+
+	return &ocpRouteSource{
+		client:                   ocpClient,
+		namespace:                namespace,
+		annotationFilter:         annotationFilter,
+		fqdnTemplate:             tmpl,
+		combineFQDNAnnotation:    combineFQDNAnnotation,
+		ignoreHostnameAnnotation: ignoreHostnameAnnotation,
+		routeInformer:            routeInformer,
+	}, nil
+}
+
+func (ors *ocpRouteSource) AddEventHandler(handler func() error, stopChan <-chan struct{}, minInterval time.Duration) {
+}
+
+// Endpoints returns endpoint objects for each host-target combination that should be processed.
+// Retrieves all OpenShift Route resources on all namespaces
+func (ors *ocpRouteSource) Endpoints() ([]*endpoint.Endpoint, error) {
+	ocpRoutes, err := ors.routeInformer.Lister().Routes(ors.namespace).List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	ocpRoutes, err = ors.filterByAnnotations(ocpRoutes)
+	if err != nil {
+		return nil, err
+	}
+
+	endpoints := []*endpoint.Endpoint{}
+
+	for _, ocpRoute := range ocpRoutes {
+		// Check controller annotation to see if we are responsible.
+		controller, ok := ocpRoute.Annotations[controllerAnnotationKey]
+		if ok && controller != controllerAnnotationValue {
+			log.Debugf("Skipping OpenShift Route %s/%s because controller value does not match, found: %s, required: %s",
+				ocpRoute.Namespace, ocpRoute.Name, controller, controllerAnnotationValue)
+			continue
+		}
+
+		orEndpoints := endpointsFromOcpRoute(ocpRoute, ors.ignoreHostnameAnnotation)
+
+		// apply template if host is missing on OpenShift Route
+		if (ors.combineFQDNAnnotation || len(orEndpoints) ==0) && ors.fqdnTemplate != nil {
```

Comment:
nit: add a space after the `==`. Should be `en(orEndpoints) == 0`.

Reply:
easy peasy :)